File: newgame.rb
#!d:\ruby193\bin\ruby
# -*- encoding: utf-8 -*-

#!/usr/bin/ruby

require './common_ui.rb'
require "./userinfo.rb"

#
# 対局登録画面
#
def newgame_screen(header, title, name, userinfo)
  scriptname = File.basename($0)

  blogin = (userinfo.user_id != nil && userinfo.user_id != "")

  CommonUI::HTMLHead(header, title)

  if blogin
    CommonUI::HTMLmenuLogIn(name)
  else
    CommonUI::HTMLmenu(name)
  end

  print <<-STYLESHEET
<style type="text/css">
<!--
  table { font-size: 2rem; }
  input { font-size: 2rem; }
  button { font-size: 2rem; }
-->
</style>
STYLESHEET

  print <<-JAVASCRIPT
<script type='text/javascript'>
<!--
function bytes2(str) {
    return encodeURIComponent(str).replace(/%../g, "x").length;
}

function validatemail(str) {
    return /^[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+@[a-zA-Z0-9-]+(?:\.[a-zA-Z0-9-]+)*$/.test(str);
}

function check_form()
{
  var nmismatch = 0;
  var alertmsg = '';

  var name;
  name = document.getElementById('rname');
  if (bytes2(name.value) < 4) {
    document.getElementById('player1').style.backgroundColor = 'red';
    document.getElementById('tremail').style.backgroundColor = 'red';
    alertmsg += 'name is too short in player1!\\n';
    ++nmismatch;
  } else {
    document.getElementById('player1').style.backgroundColor = 'transparent';
    document.getElementById('tremail').style.backgroundColor = 'transparent';
  }

  name = document.getElementById('rname2');
  if (bytes2(name.value) < 4) {
    document.getElementById('player2').style.backgroundColor = 'red';
    document.getElementById('tremail2').style.backgroundColor = 'red';
    alertmsg += 'name is too short  in player2!\\n';
    ++nmismatch;
  } else {
    document.getElementById('player2').style.backgroundColor = 'transparent';
    document.getElementById('tremail2').style.backgroundColor = 'transparent';
  }

  var email = document.getElementById('remail');
  if (validatemail(email.value)) {
  } else {
    document.getElementById('player1').style.backgroundColor = 'red';
    document.getElementById('tremail').style.backgroundColor = 'red';
    alertmsg += 'e-mail addresses is strange in player1!\\n';
    ++nmismatch;
  }

  email = document.getElementById('remail2');
  if (validatemail(email.value)) {
  } else {
    document.getElementById('player2').style.backgroundColor = 'red';
    document.getElementById('tremail2').style.backgroundColor = 'red';
    alertmsg += 'e-mail addresses is strange in player2!\\n';
    ++nmismatch;
  }

  if (nmismatch === 0) {
    document.forms['gennewgame'].submit();
  } else {
    window.alert(alertmsg);
  }
}

function pre_check() {
  var name = document.getElementById('rname');
  var name2 = document.getElementById('rname2');
  var email = document.getElementById('remail');
  var email2 = document.getElementById('remail2');

  var postmsg = "rname="+name.value+"&rname2="+name2.value;
  postmsg += "&remail="+email.value+"&remail2="+email2.value;

  var ajax = new XMLHttpRequest();
  if (ajax != null) {
    var komanim = document.getElementById('komanim');
    komanim.style.display = 'inline';
    var btn = document.getElementById('precheck')
    btn.disabled = true;

    ajax.open('POST', 'checknewgame.rb', true);
    ajax.overrideMimeType('text/plain; charset=UTF-8');
    ajax.send(postmsg);

    ajax.onreadystatechange = function () {
    	switch (ajax.readyState) {
    	case 4:
        var komanim = document.getElementById('komanim');
        komanim.style.display = 'none';
      var btn = document.getElementById('precheck')
      btn.disabled = false;
        var status = ajax.status;
    		if (status == 0) {  // XHR 通信失敗
    			alert("XHR 通信失敗");
    		} else {  // XHR 通信成功
    			if ((200 <= status && status < 300) || status == 304) {
            // リクエスト成功
            window.alert(ajax.responseText);
    			} else {  // リクエスト失敗
    				alert("その他の応答:" + status);
    			}
    		}
    		break;
    	}
    };
    ajax.onload = function(e) {
      utf8text = ajax.responseText;
    };
  }
}

function lets_furigoma() {
  var koma = document.getElementById('furikomanim1');
  koma.style.display = 'inline';
  koma = document.getElementById('furikomafu1');
  koma.style.display = 'none';
  koma = document.getElementById('furikomato1');
  koma.style.display = 'none';
  koma = document.getElementById('furikomanim2');
  koma.style.display = 'inline';
  koma = document.getElementById('furikomafu2');
  koma.style.display = 'none';
  koma = document.getElementById('furikomato2');
  koma.style.display = 'none';
  koma = document.getElementById('furikomanim3');
  koma.style.display = 'inline';
  koma = document.getElementById('furikomafu3');
  koma.style.display = 'none';
  koma = document.getElementById('furikomato3');
  koma.style.display = 'none';
  koma = document.getElementById('furikomanim4');
  koma.style.display = 'inline';
  koma = document.getElementById('furikomafu4');
  koma.style.display = 'none';
  koma = document.getElementById('furikomato4');
  koma.style.display = 'none';
  koma = document.getElementById('furikomanim5');
  koma.style.display = 'inline';
  koma = document.getElementById('furikomafu5');
  koma.style.display = 'none';
  koma = document.getElementById('furikomato5');
  koma.style.display = 'none';

  var btn = document.getElementById('btnfurigoma')
  btn.disabled = true;
  koma = document.getElementById('furigoma');
  koma.value = "";
  setTimeout("ontimer_furigoma()", 1000);
}
function ontimer_furigoma() {
  var furikoma = document.getElementById('furigoma');
  var btn = document.getElementById('btnfurigoma')
  var koma;
  var komame = furikoma.value.length
  if (komame == 0) {
    koma = document.getElementById('furikomanim1');
    koma.style.display = 'none';
    if (Math.random() < 0.5) {
      furikoma.value += "F";
      koma = document.getElementById('furikomafu1');
    } else {
      furikoma.value += "T";
      koma = document.getElementById('furikomato1');
    }
    koma.style.display = 'inline';
    setTimeout("ontimer_furigoma()", 1000);
  } else if (komame == 1) {
    koma = document.getElementById('furikomanim2');
    koma.style.display = 'none';
    if (Math.random() < 0.5) {
      furikoma.value += "F";
      koma = document.getElementById('furikomafu2');
    } else {
      furikoma.value += "T";
      koma = document.getElementById('furikomato2');
    }
    koma.style.display = 'inline';
    setTimeout("ontimer_furigoma()", 1000);
  } else if (komame == 2) {
    koma = document.getElementById('furikomanim3');
    koma.style.display = 'none';
    if (Math.random() < 0.5) {
      furikoma.value += "F";
      koma = document.getElementById('furikomafu3');
    } else {
      furikoma.value += "T";
      koma = document.getElementById('furikomato3');
    }
    koma.style.display = 'inline';
    setTimeout("ontimer_furigoma()", 1000);
  } else if (komame == 3) {
    koma = document.getElementById('furikomanim4');
    koma.style.display = 'none';
    if (Math.random() < 0.5) {
      furikoma.value += "F";
      koma = document.getElementById('furikomafu4');
    } else {
      furikoma.value += "T";
      koma = document.getElementById('furikomato4');
    }
    koma.style.display = 'inline';
    setTimeout("ontimer_furigoma()", 1000);
  } else if (komame == 4) {
    koma = document.getElementById('furikomanim5');
    koma.style.display = 'none';
    if (Math.random() < 0.5) {
      furikoma.value += "F";
      koma = document.getElementById('furikomafu5');
    } else {
      furikoma.value += "T";
      koma = document.getElementById('furikomato5');
    }
    koma.style.display = 'inline';
    btn.disabled = false;
  } else {
    btn.disabled = false;
  }
}
-->
function furifusen() {
  var name = document.getElementById('rname').value;
  var btn = document.getElementById('btnfurigoma')
  btn.value = name + 'の振り歩先で振り駒';
}
</script>
JAVASCRIPT

  print <<-FORM_NEW_GAME
<FORM action='#{scriptname}?gennewgame' method=post name='gennewgame'>
<TABLE>
 <TR id='player1'>
  <TD rowspan=2>Player 1</TD><TD>name</TD><TD><INPUT name='rname' id='rname' type=text size=25 required onKeyup="furifusen()"></TD>
 </TR>
 <TR id='tremail'>
  <TD>e-mail</TD><TD><INPUT name='remail' id='remail' type=email size=25 required></TD>
 </TR>
 <TR id='player2'>
  <TD rowspan=2>Player 2</TD><TD>name</TD><TD><INPUT name='rname2' id='rname2' type=text size=25 required></TD>
 </TR>
 <TR id='tremail2'>
  <TD>e-mail</TD><TD><INPUT name='remail2' id='remail2' type=email size=25 required></TD>
 </TR>
 <TR>
  <TD colspan=3>
  <input type='button' value='submit' onClick='check_form();'>&nbsp;
  <input type='reset'>&nbsp;
  <input type='button' id='precheck' value='pre-check' onClick='pre_check();'>
  <img id='komanim' src='image/komanim.gif' style='display:none'></TD>
 </TR>
 <TR>
  <TD colspan=3>
   <input type="button" id='btnfurigoma' onClick='lets_furigoma();' value='Player1の振り歩先で振り駒'>
   <input type="hidden" id="furigoma" name="furigoma" value="FTFTF">
   <img id='furikomanim1' src='image/komanim.gif' style='display:none' width='32' height='32'>
   <img id='furikomafu1' src='image/komafu.png' style='display:none' width='32' height='32'>
   <img id='furikomato1' src='image/komato.png' style='display:none' width='32' height='32'>
   <img id='furikomanim2' src='image/komanim.gif' style='display:none' width='32' height='32'>
   <img id='furikomafu2' src='image/komafu.png' style='display:none' width='32' height='32'>
   <img id='furikomato2' src='image/komato.png' style='display:none' width='32' height='32'>
   <img id='furikomanim3' src='image/komanim.gif' style='display:none' width='32' height='32'>
   <img id='furikomafu3' src='image/komafu.png' style='display:none' width='32' height='32'>
   <img id='furikomato3' src='image/komato.png' style='display:none' width='32' height='32'>
   <img id='furikomanim4' src='image/komanim.gif' style='display:none' width='32' height='32'>
   <img id='furikomafu4' src='image/komafu.png' style='display:none' width='32' height='32'>
   <img id='furikomato4' src='image/komato.png' style='display:none' width='32' height='32'>
   <img id='furikomanim5' src='image/komanim.gif' style='display:none' width='32' height='32'>
   <img id='furikomafu5' src='image/komafu.png' style='display:none' width='32' height='32'>
   <img id='furikomato5' src='image/komato.png' style='display:none' width='32' height='32'>
  </TD>
 </TR>
</TABLE>
</FORM>
FORM_NEW_GAME
  CommonUI::HTMLfoot()
end

File: js/newgame.js
/* - - - - - - - - - - */
/* common in this file */
/* - - - - - - - - - - */

function bytes2(str) {
  return encodeURIComponent(str).replace(/%../g, "x").length;
}

function validatemail(str) {
  return /^[\w.!#$%&'*+/=?^`{|}~-]+@[a-zA-Z\d-]+(?:\.[a-zA-Z\d-]+)*$/.test(str);
}

/* - - - - - - */
/* new game 1  */
/* - - - - - - */

function chgbgcolor(bOK, id_ply, id_eml)
{
  var ply = document.getElementById(id_ply).style;
  var eml = document.getElementById(id_eml).style;
  if (bOK) {
    ply.backgroundColor = 'transparent';
    eml.backgroundColor = 'transparent';
  } else {
    ply.backgroundColor = 'red';
    eml.backgroundColor = 'red';
  }
}

function check_name(id_name, id_ply, id_eml)
{
  var name = document.getElementById(id_name);
  var ret = (bytes2(name.value) >= 4);

  chgbgcolor(ret, id_ply, id_eml);

  return ret;
}

function check_email(id_email, id_ply, id_eml)
{
  var email = document.getElementById(id_email);
  var ret = validatemail(email.value);

  chgbgcolor(ret, id_ply, id_eml);

  return ret;
}

function puterr_or_submit(ng, form, eid, msg)
{
  if (ng === 0) {
    document.forms[form].submit();
  } else {
    document.getElementById(eid).innerText = msg;
  }
}

function check_form()
{
  var nmismatch = 0;
  var alertmsg = '';

  if (!check_name('rname', 'player1', 'tremail')) {
    alertmsg += 'name is too short in player1!\n';
    ++nmismatch;
  }
  if (!check_name('rname2', 'player2', 'tremail2')) {
    alertmsg += 'name is too short in player2!\n';
    ++nmismatch;
  }

  if (!check_email('remail', 'player1', 'tremail')) {
    alertmsg += 'e-mail addresses is strange in player1!\n';
    ++nmismatch;
  }
  if (!check_email('remail2', 'player2', 'tremail2')) {
    alertmsg += 'e-mail addresses is strange in player2!\n';
    ++nmismatch;
  }

  puterr_or_submit(nmismatch, 'gennewgame', 'errmsg', alertmsg);
}

function precheck_result(status, resp)
{
  var komanim = document.getElementById('komanim');
  komanim.style.display = 'none';
  var btn = document.getElementById('precheck');
  btn.disabled = false;
  var errmsg = document.getElementById('errmsg');
  if (status === 0) {  // XHR 通信失敗
    errmsg.innerText = "XHR 通信失敗";
  } else if ((200 <= status && status < 300) || status === 304) {
    // XHR 通信成功, リクエスト成功
    errmsg.innerText = resp;
  } else {
    // XHR 通信成功, リクエスト失敗
    errmsg.innerText = "その他の応答:" + status;
  }
}

function pre_check() {
  var name = document.getElementById('rname');
  var name2 = document.getElementById('rname2');
  var email = document.getElementById('remail');
  var email2 = document.getElementById('remail2');

  var postmsg = "rname="+name.value+"&rname2="+name2.value;
  postmsg += "&remail="+email.value+"&remail2="+email2.value;

  var ajax = new XMLHttpRequest();
  if (ajax === null)
    return;

  var komanim = document.getElementById('komanim');
  komanim.style.display = 'inline';
  var btn = document.getElementById('precheck')
  btn.disabled = true;

  ajax.open('POST', './index.rb?checknewgame', true);
  ajax.overrideMimeType('text/plain; charset=UTF-8');
  ajax.send(postmsg);

  ajax.onreadystatechange = function () {
    switch (ajax.readyState) {
    case 4:
      precheck_result(ajax.status, ajax.responseText);
      break;
    }
  };
}

function init_furigoma(sub)
{
  var koma, id;
  for (var i = 1; i <= 5; ++i) {
    id = 'furikomanim' + sub + i;
    koma = document.getElementById(id);
    koma.style.display = 'inline';
    id = 'furikomafu' + sub + i;
    koma = document.getElementById(id);
    koma.style.display = 'none';
    id = 'furikomato' + sub + i;
    koma = document.getElementById(id);
    koma.style.display = 'none';
  }
}

function lets_furigoma() {
  init_furigoma('')

  var btn = document.getElementById('btnfurigoma')
  btn.disabled = true;
  var koma = document.getElementById('furigoma');
  koma.value = "";
  setTimeout(function() {
    ontimer_furigoma()
  }, 1000);
}

function randomchoose(anim, fu, to) {
  var koma = document.getElementById(anim);
  koma.style.display = 'none';
  var value;
  if (Math.random() < 0.5) {
    value = "F";
    koma = document.getElementById(fu);
  } else {
    value = "T";
    koma = document.getElementById(to);
  }
  koma.style.display = 'inline';
  return value;
}

function ontimer_furigoma() {
  var furikoma = document.getElementById('furigoma');
  var komame = furikoma.value.length+1;
  if (komame <= 5) {
    var idanim = 'furikomanim' + komame;
    var idfu = 'furikomafu' + komame;
    var idto = 'furikomato' + komame;
    furikoma.value += randomchoose(idanim, idfu, idto);
  }
  if (komame <= 4) {
    setTimeout(function() {
      ontimer_furigoma();
    }, 1000);
  } else {
    var btn = document.getElementById('btnfurigoma');
    btn.disabled = false;
  }
}

/* - - - - - - - - */
/* - new game 2  - */
/* - - - - - - - - */

function check_selectindex(rid, eid)
{
  var index = document.getElementById(rid).selectedIndex;
  var ret = (index >= 1);
  if (ret) {
    document.getElementById(eid).style.backgroundColor = 'transparent';
  } else {
    document.getElementById(eid).style.backgroundColor = 'red';
  }
  return ret;
}

function check_form2()
{
  var nmismatch = 0;
  var alertmsg = '';

  if (!check_selectindex('rid', 'player21')) {
    alertmsg += 'please select player1!\n';
    ++nmismatch;
  }
  if (!check_selectindex('rid2', 'player22')) {
    alertmsg += 'please select player2!\n';
    ++nmismatch;
  }

  puterr_or_submit(nmismatch, 'gennewgame2', 'errmsg2', alertmsg);
}

function lets_furigoma2() {
  init_furigoma('2')

  var btn = document.getElementById('btnfurigoma2');
  btn.disabled = true;
  var koma = document.getElementById('furigoma2');
  koma.value = "";
  setTimeout(function() {
    ontimer_furigoma2()
  }, 1000);
}

function ontimer_furigoma2() {
  var furikoma = document.getElementById('furigoma2');
  var komame = furikoma.value.length+1;
  if (komame <= 5) {
    var idanim = 'furikomanim2' + komame;
    var idfu = 'furikomafu2' + komame;
    var idto = 'furikomato2' + komame;
    furikoma.value += randomchoose(idanim, idfu, idto);
  }
  if (komame <= 4) {
    setTimeout(function() {
      ontimer_furigoma2();
    }, 1000);
  } else {
    var btn = document.getElementById('btnfurigoma2');
    btn.disabled = false;
  }
}

function furifusen() {
  var name = document.getElementById('rname').value;
  var btn = document.getElementById('btnfurigoma')
  btn.value = name + 'の振り歩先で振り駒';
}

function furifusen2() {
  var index = document.getElementById('rid').selectedIndex;
  var options = document.getElementById('rid').options;
  var text = options[index].label;
  var cap = /(.+)\(/g.exec(text);
  var btn = document.getElementById('btnfurigoma2')
  btn.value = cap[1] + 'の振り歩先で振り駒';
}
